src/lib/idp/verifyAccessToken.ts

```diff
@@ -15,50 +16,51 @@ export type VerifiedAccessToken = {
 const REDIS_INTROSPECTION_RESULT_KEY_PREFIX: string = "api:auth";
 const CACHE_EXPIRY_DELAY_IN_SECONDS: number = 300;
 
-export async function verifyAccessToken(
-  accessToken: string,
-): Promise<VerifiedAccessToken | undefined> {
+export class AccessTokenExpiredError extends Error {
+  constructor() {
+    super("Access token has expired");
+    this.name = "AccessTokenExpiredError";
+  }
+}
+
+export class AccessTokenInvalidError extends Error {
+  constructor() {
+    super("Access token is invalid");
+    this.name = "AccessTokenInvalidError";
+  }
+}
+
+export class AccessControlError extends Error {
+  constructor() {
+    super("Access is forbidden");
+    this.name = "AccessForbiddenError";
+  }
+}
+
+export async function verifyAccessToken(accessToken: string, formId: string) {
   try {
-    const cachedVerifiedAccessToken =
-      await getVerifiedAccessTokenFromCache(accessToken);
-
-    if (cachedVerifiedAccessToken !== undefined) {
-      return cachedVerifiedAccessToken;
-    }
-
-    const accessTokenIntrospectionResult =
-      await introspectAccessToken(accessToken);
-
-    if (accessTokenIntrospectionResult.active === false) {
-      return undefined;
-    }
-
-    if (
-      accessTokenIntrospectionResult.exp === undefined ||
-      accessTokenIntrospectionResult.sub === undefined ||
-      accessTokenIntrospectionResult.username === undefined
-    ) {
-      throw new Error(
-        "Access token introspection result is missing required properties",
-      );
-    }
-
-    const verifiedAccessToken: VerifiedAccessToken = {
-      expirationEpochTime: accessTokenIntrospectionResult.exp,
-      serviceAccountId: accessTokenIntrospectionResult.sub,
-      serviceUserId: accessTokenIntrospectionResult.username,
-    };
-
-    await cacheVerifiedAccessToken(accessToken, verifiedAccessToken);
-
-    return verifiedAccessToken;
+    // Get token from cache, it if doesn't exist introspect from IDP
+    const introspectedAccessToken = await getIntrospectedAccessTokenFromCache(
+      accessToken,
+    ).then((cachedToken) => {
+      if (cachedToken !== undefined) {
+        return cachedToken;
+      }
+      return generateIntrospectedAccessToken(accessToken, formId);
+    });
+
+    // Checks for expiry and access control
+    validateIntrospectedToken(introspectedAccessToken, formId);
+
+    await cacheVerifiedAccessToken(accessToken, introspectedAccessToken);
```

Comment:
With this refactoring we are now caching the verified access token every time we process a request. This creates an extra call to Redis but also continuously extend the 5 minute expiry date (in cache) for that verified token. 
I guess this would be a problem if we decide to revoke an API key because, depending on the access token age, a user could potentially still be able to use it for up to 30 minutes.

Reply:
You're right good catch.  I'll move the cache call.